src/main/kotlin/no/nav/meldeplikt/meldekortservice/database/Queries.kt

```diff
@@ -1,8 +1,21 @@
 package no.nav.meldeplikt.meldekortservice.database
 
+import com.fasterxml.jackson.databind.ObjectMapper
```

Comment:
Det blir vel et problem med concurrency når vi kjører flere pod'er og alle leser/skriver til samme tabell.
For å unngå problemer så bør vi låse radene i MIDLERTIDIG_LAGREDE_JOURNALPOSTER som plukkes ut til behandling / hoppe over rader som allerede er låst.

I Spring har vi løst tilsvarende problem slik:
```
	String SKIP_LOCKED = "-2";

	/**
	 * Finner og låser alle hendelser som skal rekjøres. Dette er hendelser med retry-status der retry-tidspunktet er nådd
	 * (eller har passert).
	 * <p>
	 * For å unngå samtidighetskonflikter benyttes en låsemekanisme:
	 * <ul>
	 * <li>Låsemodus PESSIMISTIC_WRITE - gir eksklusiv lås på radene som leses opp.</li>
	 * <li>Hint SKIP_LOCKED - hopper over rader som allerede er låst.</li>
	 * <li>Hint FOLLOW_ON_LOCKING - benyttes fordi Oracle oppfører seg litt annereledes enn en del andre databaser: <i>Oracle
	 * uses MVCC (Multiversion Concurrency Control) so Readers don't block Writers and Writers don't block Readers. Even if you
	 * acquire a row-level lock with Oracle, and you modify that row without committing, other transactions can still read the
	 * last committed value</i>.</li>
	 * </ul>
	 *
	 * @return En liste med alle hendelser som skal rekjøres.
	 */
	@QueryHints(value = { @QueryHint(name = "javax.persistence.lock.timeout", value = SKIP_LOCKED),
			@QueryHint(name = org.hibernate.jpa.QueryHints.HINT_FOLLOW_ON_LOCKING, value = "false") })
	@Lock(LockModeType.PESSIMISTIC_WRITE)
	List<Meldekorthendelse> findByStatusAndRetryTidspunktLessThanEqualOrderByIdAsc(String status, LocalDateTime tidspunkt);
```
Det er vel mulig å lage noe tilsvarende her.

Reply:
Jeg har endret koden slik at vi har "select for update". I tillegg har vi et nytt felt process_started (timestamp).

Logikken er:
1. Vi tar alle rader som ikke har process_started (det er NULL der) eller process_started er gammel nok (mer enn 1 time gammel). Dvs. alle rader som ikke er under prosessering nå, eller som kanskje var under prosessering, men noe gikk galt med poden.
Samtidig blokkerer vi disse radene, andre prosesser kan ikke lese disse radene da.
2. Vi markerer disse radene, dvs. skriver inn current_timestamp inn i process_started.
3. Vi avslutter transaksjonen, andre prosesser kan lese disse radene igjen, men siden de er allerede markert med process_started, skal ikke disse radene hentes av andre prosesser.
4. Vi går gjennom en liste med radene og prøver å sende data fra hver rad på nytt. Hvis vi kan sende data, sletter vi raden. Hvis vi ikke kan sende data, sletter vi process_started, dvs. raden er igjen tilgjengelig for å bli prosessert senere.

Reply:
Hvis du bruker SELECT .. FOR UPDATE SKIP LOCKED; så vil du kun velge rader som ikke allerede er valgt av en annen prosess  uten noe ekstra logikk.

Reply:
Nei, kan ikke dessverre. Det er forskjellige transaksjoner:
Transaksjon 1: select for update (locked)
Prøver å sende (transaksjonen 1 er allerede avsluttet her og vi har ikke lock)
Transaksjon 2: slett eller update retries

Dvs. andre prosesser kan hente rader mellom disse 2 transaksjonene hvis vi ikke markerer rader på en eller annen måte

Reply:
hm, skjønner ikke helt.. Hvorfor er det forskjellige transaksjoner?

Reply:
Ok, jeg har endret koden slik at nå bruker vi SELECT FOR UPDATE SKIP LOCKED. Men jeg kan ikke teste dette i enhetstester fordi H2 database ikke støtter SKIP LOCKED